lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,506 @@
+/// data receipt, as opposed to originally signed value. TODO: TBD
+const TIMESTAMP_RETROGRESSION_DELTA: u32 = 7 * 24 * 3600;
+
+/// Update network graph from binary data
+///
+/// `network_graph`: network graph to be updated
+///
+/// `update_data`: `&[u8]` binary stream that comprises the update data
+pub fn update_network_graph(
+	network_graph: &network_graph::NetworkGraph,
+	update_data: &[u8],
+) -> Result<(), GraphSyncError> {
+	let mut read_cursor = io::Cursor::new(update_data);
+	read_network_graph(&network_graph, &mut read_cursor)
+}
+
+pub(crate) fn read_network_graph<R: Read>(
+	network_graph: &network_graph::NetworkGraph,
+	read_cursor: &mut R,
+) -> Result<(), GraphSyncError> {
+	let mut prefix = Vec::with_capacity(4);
+	prefix.resize(4, 0);
+	read_cursor.read_exact(&mut prefix)?;
+
+	let _deserialize_unsigned = match &prefix[..] {
+		&[76, 68, 75, 1] => true,
+		_ => {
+			return Err(GraphSyncError::ProcessingError(
+				"Prefix must equal 'LDK' and the byte 2".to_string(),
+			));
+		}
+	};
+
+	// let chain_source: Option<Box<dyn lightning::chain::Access>> = None;
+	let current_timestamp = SystemTime::now()
+		.duration_since(UNIX_EPOCH)
+		.unwrap()
+		.as_secs() as u32;
+
+	let retrogressed_timestamp = current_timestamp - TIMESTAMP_RETROGRESSION_DELTA; // back-date updates by one week
+
+	// shadow variable for ownership release after length-restricted reads
+	let mut read_cursor = read_cursor;
+
+	let chain_hash: BlockHash = Readable::read(read_cursor)?;
+
+	let announcement_count: u32 = Readable::read(read_cursor)?;
+	for _ in 0..announcement_count {
+		let features = Readable::read(read_cursor)?;
+		let short_channel_id = Readable::read(read_cursor)?;
+		let node_id_1 = Readable::read(read_cursor)?;
+		let node_id_2 = Readable::read(read_cursor)?;
+
+		let announcement_result = network_graph.update_channel_from_announcement_fragments(
+			short_channel_id,
+			retrogressed_timestamp as u64,
+			features,
+			node_id_1,
+			node_id_2
+		);
+		if let Err(lightning_error) = announcement_result {
+			if let ErrorAction::IgnoreDuplicateGossip = lightning_error.action {
+				// everything is fine, just a duplicate channel announcement
+			} else {
+				return Err(lightning_error.into());
+			}
+		}
+	}
+
+	let update_count: u32 = Readable::read(read_cursor)?;
+	for _current_update_index in 0..update_count {
+		// println!("update index: {}", _current_update_index);
+
+		let short_channel_id: u64 = Readable::read(read_cursor)?;
+		let channel_flags: u8 = Readable::read(read_cursor)?;
+
+		// flags are always sent in full, and hence always need updating
+		let standard_channel_flags = channel_flags & 0b_0000_0011;
+
+		let synthetic_update = if channel_flags & 0b_1000_0000 == 0 {
+			// full update
+			// println!("channel flags: {:b}, scid: {}, full update", channel_flags, short_channel_id);
+
+			let cltv_expiry_delta: u16 = Readable::read(&mut read_cursor)?;
```

Comment:
Why are we reading these all here? It looks like we can just set the flags and read them later anyway, might as well optimize a hair and not bother reading the fields that have flags assigned here and instead assign them default (ie today's-most-common) values and let the server skip them if they match.

Reply:
I think I'm still missing something here. The flags are set when the value differs from the previous one, otherwise the server doesn't send the value.

Reply:
We already have `channel_flags` read outside the if block here, so we can take the reads at https://github.com/lightningdevkit/rust-lightning/blob/7ea4a63c9c68d68266ceca98eb41c7be2ad3a25f/lightning-graph-sync/src/processing.rs#L215-L241 and move them outside the if block. Then, here, in the `if channel_flags & 0b1000_000 == 0` block, assign each field a "default" that is the most common we see on the network today. That way, we can avoid reading a bunch of the fields if they're the most common field.

Reply:
I think I understand. I wonder whether it's the most efficient mechanism for incremental updates, given that here, we have the a trivariate combination of a) was the setting changed? b) was it set to the default? c) if it was not set to the default, which value was it set to? thereby requiring two bit flags instead of one.

Reply:
Hmm? when you see a "full update" treat it as "all fields were set to the default, except those with the changed flag".